rbd/mirror_nautilus.go

```diff
@@ -0,0 +1,379 @@
+//go:build nautilus
+// +build nautilus
+
+package rbd
+
+// #cgo LDFLAGS: -lrbd
+// #include <errno.h>
+// #include <stdlib.h>
+// #include <rados/librados.h>
+// #include <rbd/librbd.h>
+import "C"
+import (
+	"unsafe"
+
+	"github.com/ceph/go-ceph/rados"
+)
+
+// MirrorMode indicates the current mode of mirroring that is applied onto a
+// pool. A pool that doesn't have an explicit mirroring mode applied to it is
+// said to be disabled - that's the default.
+type MirrorMode int
+
+const (
+	// MirrorModeDisabled disables mirroring.
+	MirrorModeDisabled = MirrorMode(C.RBD_MIRROR_MODE_DISABLED)
+	// MirrorModeImage enables mirroring on a per-image basis.
+	MirrorModeImage = MirrorMode(C.RBD_MIRROR_MODE_IMAGE)
+	// MirrorModePool enables mirroring on all journaled images.
+	MirrorModePool = MirrorMode(C.RBD_MIRROR_MODE_POOL)
+)
+
+// MirrorModeGet returns the mode of mirroring currently applied to a pool.
+//
+// Note: this can only be used if go-ceph is compiled with the `nautilus` build
+// tag.
+//
+// Implements:
+//  int rbd_mirror_mode_get(rados_ioctx_t p, rbd_mirror_mode_t *mirror_mode)
+func MirrorModeGet(ioctx *rados.IOContext) (MirrorMode, error) {
+	var rmm C.rbd_mirror_mode_t
+
+	ret := C.rbd_mirror_mode_get(C.rados_ioctx_t(ioctx.Pointer()), &rmm)
+	if ret != 0 {
+		return -1, getError(ret)
+	}
+
+	return MirrorMode(rmm), nil
+}
+
+// MirrorModeSet sets the mirror mode for a pool.
+//
+// Note: this can only be used if go-ceph is compiled with the `nautilus` build
+// tag.
+//
+// Implements:
+//  rbd_mirror_mode_set(rados_ioctx_t p, rbd_mirror_mode_t mirror_mode)
+func MirrorModeSet(ioctx *rados.IOContext, mode MirrorMode) error {
+	var cMode C.rbd_mirror_mode_t = C.rbd_mirror_mode_t(mode)
+
+	ret := C.rbd_mirror_mode_set(C.rados_ioctx_t(ioctx.Pointer()), cMode)
+
```

Comment:
This doesn't check if the pointer is `nil`. Normally we write the code like this:
```go
	if err := ioctx.validate(); err != nil {
		return err
	}
	ret := C.rbd_mirror_mode_set(ioctx.ioctx, cMode)
```
if this would be in the rados package. Because this is a special case, I might not care too much, but otherwise we should probably export some more methods to make that better usable from outside. WDYT @phlogistonjohn ?

Reply:
Use the existing helper function for rbd `cephIoctx` like so:
```
   ret := C.rbd_mirror_mode_set(cephIoctx(ioctx), cMode)
```

Thats how we do it pretty much everywhere else in rados. If we want to enhance the behavior we can/will do it in that single function.